juju/juju
# refactor: move controller facade over to model services

- The `Controller` facade has become a multi-model facade.
- It now uses the following services: ModelService, ModelConfigService (for each model), AgentService
- Controller facade v11 has been removed, only v12 is supported in 4.0. The ModelConfig method no longer exists. All client methods are now using the ModelConfig facade instead (see #17732).
- Controller facade tests are slowly being ported over to using mocked services.
- The service factory now has methods `ModelAgent` which returns a model-scoped agent service, and `Agent` which returns a general agent service to access the agent version for any model. A `ControllerAgentVersion` method has been added for convenience.
- The ModelStatus api client decodes its errors using `params.TranslateWellKnownError`.
- Added a `CreateTestModelWithConfig` method to control parameters for created test models.

## Future work

- In #17841, we will add new model service methods that this facade will use to retrieve desired model information in the `AllModels` and `HostedModelConfigs` facade methods.

## Checklist




## QA steps



The following code is affected:
- facade methods: `DashboardConnectionInfo`, `AllModels`, `ListBlockedModels`, `HostedModelConfigs`, `InitiateMigration`
- commands: `dashboard`, `destroy-controller`, `kill-controller`, `controllers`, `show-controller`, `disabled-commands`, `migrate`.

This QA should cover all the changes:
- Run the `dashboard` Bash test suite on IAAS.
- Deploy a controller with a few models and check that `show-controller` displays all of them. Try adding models to the controller as a different user as well.
- Disable commands on some of these models, run `juju disabled-commands --all` and check all disabled commands show up.
- Test that `kill-controller` works (with hosted models, and test it with `-t 0`).
- Test a model migration from 3.6 (3620f9e or later, must support Controller facade v12)

## Links



**Jira card:** JUJU-6219
